linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CustomJSHelperBuilder.scala

```diff
@@ -0,0 +1,317 @@
+ *  }
+ *
+ *  // Call the helper, which consumes the declared inputs and leaves the result on the stack
+ *  fb += wa.Call(helperID)
+ *  }}}
+ *
+ *  The subclass [[CustomJSHelperBuilder.WithTreeEval]] provides even richer
+ *  capabilities, by allowing to evaluate arbitrary IR `Tree`s as inputs. The
+ *  builder will optimize the evaluation and transfer from Wasm to JS:
+ *
+ *  - Evaluate some trees entirely on the JS side instead of on the Wasm side.
+ *    This is notably the case for string literals, which commonly appear as
+ *    method names, and global refs/external imports, which commonly appear as
+ *    method receivers.
+ *  - For other trees, evaluate them on the Wasm side, but fuse their boxing
+ *    operation with the Wasm-to-JS interoperability layer, by choosing the
+ *    best possible Wasm type for the helper function.
+ *
+ *  When using the `WithTreeEval` subclass, the user must provide a concrete
+ *  method to evaluate `Tree`s at call site with a given expected type. In the
+ *  context of `FunctionEmitter`, that directly translates to the `genTree`
+ *  method.
+ *
+ *  A typical scenario looks like:
+ *
+ *  {{{
+ *  val builder = new CustomJSHelperBuilder.WithTreeEval() {
+ *    protected def evalTreeAtCallSite(tree: Tree, expectedType: Type): Unit = ???
+ *  }
+ *
+ *  // Register sources of JS global refs that we will need to avoid clashes with
+ *  builder.registerForGlobals(...) // Trees, JSNativeLoadSpecs, etc.
+ *  ...
+ *
+ *  // Add inputs that will be evaluated on the JS side if possible, otherwise
+ *  // evaluated on the Wasm call site, and whose values are then available as
+ *  // `js.Tree`s in the JS helper:
+ *  val xRef = builder.addInput(someIRTree)
+ *  ...
+ *  }}}
+ *
+ *  In addition to the input management, `CustomJSHelperBuilder` provides tools
+ *  for local name management, and generation of `js.ParamDef`s for inner
+ *  functions.
+ */
+class CustomJSHelperBuilder()(implicit ctx: WasmContext, pos: Position) {
+  import CustomJSHelperBuilder._
+
+  private val registeredGlobalRefs = mutable.Set.empty[String]
+  private val allocatedLocalIdents = mutable.Set.empty[String]
+
+  private val jsParamDefs = mutable.ListBuffer.empty[js.ParamDef]
+  private val wasmParamTypes = mutable.ListBuffer.empty[watpe.Type]
+
+  def freshLocalIdent(origName: String): js.Ident = {
+    var result = origName
+    var index = 0
+    while (registeredGlobalRefs.contains(result) || !allocatedLocalIdents.add(result)) {
+      index += 1
+      result = origName + index
+    }
+    js.Ident(result)
+  }
+
+  def freshLocalIdent(name: LocalName): js.Ident =
+    freshLocalIdent(ctx.jsNameGen.genName(name))
+
+  private def addParamDef(origName: String, wasmType: watpe.Type): js.VarRef = {
+    val jsParamDef = js.ParamDef(freshLocalIdent(origName))
+    jsParamDefs += jsParamDef
+    wasmParamTypes += wasmType
+    jsParamDef.ref
+  }
+
+  def registerGlobalRef(name: String): Unit = {
```

Comment:
It is a pity that we need the registrations. For more complex use cases (notably JS class creation), I'm a bit concerned that it is quite brittle: incorrect registration is unlikely to be caught by tests, since it highly depends on the specific user code. I don't think it is realistic to design our test suite to test all such code paths.

From what I understand, even making generated refs lazy is not enough without impacting the control flow of the generating side (but maybe it would be acceptable, currently all of them *have* to mention inputs before any codegen to register, so they might as well allocate input refs there and back-refer; might need some additional zipping, but maybe nothing too bad).

Making refs virtual is probably not an option :P

Reply:
Huh. The laziness didn't cross my mind. 

I think we could generalize our use of `DelayedIdent`s to also be accepted for local idents. Then we can return the delayed idents here, and allocate them in `build`. The call sites wouldn't see the difference!